internal/datastore/crdb/tx.go

```diff
@@ -37,36 +50,80 @@ type transactionFn func(tx pgx.Tx) error
 
 type executeTxRetryFunc func(context.Context, conn, pgx.TxOptions, transactionFn) error
 
+// RetryError wraps an error that prevented the transaction function from being retried.
+type RetryError struct {
+	err error
+}
+
+// Error returns the wrapped error
+func (e RetryError) Error() string {
+	return fmt.Errorf("unable to retry conflicted transaction: %w", e.err).Error()
+}
+
+// Unwrap returns the wrapped, non-retriable error
+func (e RetryError) Unwrap() error {
+	return e.err
+}
+
 func executeWithMaxRetries(max int) executeTxRetryFunc {
 	return func(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transactionFn) (err error) {
-		return execute(ctx, conn, txOptions, fn, max)
+		return executeWithResets(ctx, conn, txOptions, fn, max)
 	}
 }
 
-// adapted from https://github.com/cockroachdb/cockroach-go
-func execute(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transactionFn, maxRetries int) (err error) {
+// executeWithResets executes transactionFn and resets the tx when ambiguous crdb errors are encountered.
+func executeWithResets(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transactionFn, maxRetries int) (err error) {
+	var retries, resets int
+	defer func() {
+		retryHistogram.Observe(float64(retries))
+		resetHistogram.Observe(float64(resets))
+	}()
+
 	var tx pgx.Tx
-	tx, err = conn.BeginTx(ctx, txOptions)
-	if err != nil {
-		return err
-	}
 	defer func() {
 		if err == nil {
-			_ = tx.Commit(ctx)
-			return
+			if commitErr := tx.Commit(ctx); commitErr == nil {
+				return
+			}
+			log.Ctx(ctx).Err(err).Msg("failed tx commit")
+		}
+
+		if rollbackErr := tx.Rollback(ctx); rollbackErr != nil {
+			log.Ctx(ctx).Err(err).Msg("error during tx rollback")
```

Comment:
should we log the `rollbackErr` (and `commitErr` above) too? not sure I've ever seen one so I don't know if they're useful or not